### DISCUSSION TITLE
solver performance on regression test suite

### DISCUSSION THREAD

**Discussion by fclaudel** [bitwuzla/bitwuzla#143] [General]
Hi,
I've run the smt2 test files from Bitwuzla's regression testsuite in test/regress/rewrite/bv  through release binaries of bitwuzla and yices_smt2 binary (https://yices.csl.sri.com). I've updated the binaries to use both a monotonic clock with the stat displayed in microseconds, so as to get more precise and comparable results. Timings are averaged on 100 runs per testcase and the same SAT or UNSAT status must match between the solvers for the test to be considered valid.

The outcome is unexpected, with some wild differences between them. On both extremes are these 2:
RESULT: test/regress/solver/bv/div.btor.smt2 108 16409  gives 108μs for yices-smt2 and 16409μs for Bitwuzla
RESULT: test/regress/rewrite/bv/rw24.btor.smt2 102260 201 gives 102260μs for yices-smt2 and 201μs for Bitwuzla

So, depending on the testcases, the performance can vary drastically. Can you give some hindsight as to why such differences can be seen? Is there some way to configure Bitwuzla or change the query to get closer to yices behaviour for some kinds of queries frequently used e.g. division (with the tradeoff of slower queries on other presumably)?

**Comment by mpreiner** [comment]
For `test/regress/solver/bv/div.btor.smt2`, Yices2 seems to solve it without bit-blasting (and that's why it is faster), whereas for `test/regress/rewrite/bv/rw24.btor.smt2` Bitwuzla rewrites the formula to false (and Yices2 may need to bit-blast).

**Comment by aniemetz** [comment]
Using abstraction (option `--abstraction`) on this input signifcantly reduces solve time (should be comparable with Yices' time, I didn't set up with @fclaudel timer configuration).

**Comment by fclaudel** [comment]
Many thanks for your answers.

About setup, I used yices 2.6.2 release for which I patched src/frontend/yices_smt2.c to measure microseconds elapsed in the loop that calls parse_smt2_command, using 2 calls to clock_gettime (CLOCK_MONOTONIC, ..)
I used bitwuzla 0.6.0 release for which I patched src/util/statistics.cpp to use chrono::microseconds instead of chrono::milliseconds and extracted duration reported by solving_context::time_solve.

I've now modified the script to pass --abstraction and get these timings for Bitwuzla:
test/regress/solver/bv/div.btor.smt2 446
test/regress/rewrite/bv/rw24.btor.smt2 137

So the option does make a difference for div.btor.smt2. Can you tell what it does and why it is disabled by default (defaults to 0 in the C++ API)? Is there another option that can reduce the time difference further for the testcase?

**Reply by aniemetz** [reply]
This enables our new abstraction module for abstracting bit-vector arithmetic operators, see [Aina Niemetz, Mathias Preiner and Yoni Zohar. Scalable Bit-Blasting with Abstractions. CAV 2024, Springer, 2024.](https://doi.org/10.1007/978-3-031-65627-9_9). The minimum bit-width of relevant terms to abstract can be configured via option --abstraction-bv-size.

We plan to enable it by default in a future release.

**Reply by aniemetz** [reply]
Also, thanks for testing this with your setup. 446 for div.btor.smt2 matches what I observed and is great, considering that this actually involves firing off a solving procedure (it isn't solved via rewriting as in Yices).
